components/navigable-container/stoppers.js

```diff
@@ -0,0 +1,41 @@
+/**
+ * WordPress Dependencies
+ */
+import { keycodes } from '@wordpress/utils';
+
+/**
+ * Module Constants
+ */
+const { UP, DOWN, LEFT, RIGHT, TAB } = keycodes;
+
+function stopBubbling( event ) {
+	// Prevents arrow key handlers bound to the document directly interfering
+	event.nativeEvent.stopImmediatePropagation();
+	event.stopPropagation();
+}
+
+function handleHorizontalArrows( event ) {
+	// Do not preventDefault if inside a textarea or input where the arrow keys should move inside it.
+	if ( ! [ 'textarea', 'input' ].indexOf( event.target.nodeName.toLowerCase() ) ) {
+		event.preventDefault();
+	}
+
+	stopBubbling( event );
+}
+
+function stopEvent( event ) {
+	event.preventDefault();
+	stopBubbling( event );
+}
+
+export function stopEventIfRequired( stopArrowEvents, stopTabEvents, event ) {
+	const { keyCode } = event;
+
+	if ( stopArrowEvents && [ LEFT, RIGHT ].indexOf( keyCode ) > -1 ) {
+		handleHorizontalArrows( event );
+	} else if ( stopArrowEvents && [ UP, DOWN ].indexOf( keyCode ) > -1 ) {
+		stopEvent( event );
+	} else if ( stopTabEvents && TAB === keyCode ) {
+		stopEvent( event );
+	}
+}
```

Comment:
Thinking about this more. I'm wondering why we're not systematically preventing and stopPropagation consistently. 

I mean if we're using `arrows` in a navigable menu, we expect these arrows to navigate to other elements, thus, we need to preventEverything. And if we're using `tabs` same.

By this I mean, it makes no sense to me to use arrow navigation if your children could contain inputs/textareas...

which brings us the my simple check: `preventBubblingEvents && offset !== 0`. This would work because if use arrows, the offset will be different than `0` only for arrow keys and same for tabs.

Reply:
Hmm. Interesting idea. I have no problem with this but it would mean that the person writing the offset function would need to understand that returning 0 for an offset, even if it is just because they have hit the edge, means that the event will not be handled and will bubble up (which causes problems when things like WritingFlow listen to arrow keys above). I guess given that cycle is done here (and not done by the offset calculation code which could return 0 at the edges), then the offset function will just have to be clear that you do not return 0 unless you want the event to bubble.

My personal style is to avoid conflating stopping the event and moving the focus, but I can see the benefits (ensure the two are kept in sync). This is done most easily by having a complicated function, so I can see why that is undesirable in the API. Your approach will work if you work on the assumption that offset 0 means do not bubble.

Another problem with this is that it means that for situations where you have a horizontal menu (left, right arrow keys), pressing up and down will have an offset of 0, so they will bubble to WritingFlow and cause problems. Initially, I introduced this so that even the arrow keys that weren't involved in moving the focus were being stopped. Otherwise, the other keys (up instead of left) etc. are going to bubble.

Reply:
As discussed in DM, your idea of returning `undefined` for do not stop this event, and `0` for stop it, but don't move seems good to me.